parser-typechecker/src/Unison/Runtime/Machine.hs

```diff
@@ -279,150 +279,152 @@ exec !env !denv !ustk !bstk !k (BPrim1 LOAD i) = do
      -> Stack 'UN -> Stack 'BX -> K -> Section -> IO ()
-eval !env !denv !ustk !bstk !k (Match i (TestT df cs)) = do
+eval !env !denv !trackThreads !ustk !bstk !k (Match i (TestT df cs)) = do
   t <- peekOffBi bstk i
-  eval env denv ustk bstk k $ selectTextBranch t df cs
-eval !env !denv !ustk !bstk !k (Match i br) = do
+  eval env denv trackThreads ustk bstk k $ selectTextBranch t df cs
+eval !env !denv !trackThreads !ustk !bstk !k (Match i br) = do
   n <- peekOffN ustk i
-  eval env denv ustk bstk k $ selectBranch n br
-eval !env !denv !ustk !bstk !k (Yield args)
+  eval env denv trackThreads ustk bstk k $ selectBranch n br
+eval !env !denv !trackThreads !ustk !bstk !k (Yield args)
   | asize ustk + asize bstk > 0 , BArg1 i <- args = do
-    peekOff bstk i >>= apply env denv ustk bstk k False ZArgs
+    peekOff bstk i >>= apply env denv trackThreads ustk bstk k False ZArgs
   | otherwise = do
     (ustk, bstk) <- moveArgs ustk bstk args
     ustk <- frameArgs ustk
     bstk <- frameArgs bstk
-    yield env denv ustk bstk k
-eval !env !denv !ustk !bstk !k (App ck r args) =
+    yield env denv trackThreads ustk bstk k
+eval !env !denv !trackThreads !ustk !bstk !k (App ck r args) =
   resolve env denv bstk r
-    >>= apply env denv ustk bstk k ck args
-eval !env !denv !ustk !bstk !k (Call ck n args) =
+    >>= apply env denv trackThreads ustk bstk k ck args
+eval !env !denv !trackThreads !ustk !bstk !k (Call ck n args) =
   combSection env (CIx dummyRef n 0)
-    >>= enter env denv ustk bstk k ck args
-eval !env !denv !ustk !bstk !k (Jump i args) =
-  peekOff bstk i >>= jump env denv ustk bstk k args
-eval !env !denv !ustk !bstk !k (Let nw cix) = do
+    >>= enter env denv trackThreads ustk bstk k ck args
+eval !env !denv !trackThreads !ustk !bstk !k (Jump i args) =
+  peekOff bstk i >>= jump env denv trackThreads ustk bstk k args
+eval !env !denv !trackThreads !ustk !bstk !k (Let nw cix) = do
   (ustk, ufsz, uasz) <- saveFrame ustk
   (bstk, bfsz, basz) <- saveFrame bstk
-  eval env denv ustk bstk (Push ufsz bfsz uasz basz cix k) nw
-eval !env !denv !ustk !bstk !k (Ins i nx) = do
-  (denv, ustk, bstk, k) <- exec env denv ustk bstk k i
-  eval env denv ustk bstk k nx
-eval !_   !_    !_    !_    !_ Exit = pure ()
-eval !_   !_    !_    !_    !_ (Die s) = die s
+  eval env denv trackThreads ustk bstk (Push ufsz bfsz uasz basz cix k) nw
+eval !env !denv !trackThreads !ustk !bstk !k (Ins i nx) = do
+  (denv, ustk, bstk, k) <- exec env denv trackThreads ustk bstk k i
+  eval env denv trackThreads ustk bstk k nx
+eval !_   !_    !_    !_trackThreads !_    !_ Exit = pure ()
+eval !_   !_    !_    !_trackThreads !_    !_ (Die s) = die s
 {-# noinline eval #-}
 
-forkEval :: CCache -> Closure -> IO ThreadId
-forkEval env clo
-  = forkIO (apply1 err env clo)
+forkEval :: CCache -> (ThreadId -> IO ()) -> Closure -> IO ThreadId
+forkEval env trackThreads clo
+  = do threadId <- forkIO (apply1 err env trackThreads clo)
+       trackThreads threadId
```

Comment:
Unfortunately I think this will continue to accumulate an unbounded number of threadIds in the Set (only when running in UCM though).

Threads should probably remove themselves from the cleanup pool when they finish or fail.

To accomodate that I'd probably just switch from threading through a `trackThreadId` function to threading through a `fork :: IO a -> IO ThreadId` combinator and have it handle all the appropriate book-keeping on its own,

but I'll hold off on implementing that until we decide on which approach to take.

Reply:
You could use `forkFinally` to subtract the thread id from the set when a thread exits.

I've been trying to think of a better way to cascade the thread kill messages, but I haven't come up with anything.

Reply:
Sounds good, threads now clean up after themselves 👍🏼